### PR TITLE
write actual weights to files on disk

### PR DESCRIPTION
**Pull request by eddogola**
## Description

save model parameters(params+buffers) to disk, following the huggingface format, in multiple binary files depending on world size.

Fixes #(issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by eddogola** [comment]
@kwen2501 @wz337

File: pippy/hf/_SaveModule.py [hunk 1]
```diff
@@ -81,12 +86,29 @@ def _save_index(
...
+            #  add ckpt size once
+            if old_name not in weight_map:
+                index_dict["metadata"]["total_size"] += _get_param_size(param)  # type: ignore
+
             weight_map[old_name] = binary_filename
 
-            total_size += _get_param_size(param)
+            state_dict[old_name] = param
 
+    _save_weights(state_dict, file_to_param_names, checkpoint_dir)
```
Comment (wz337) [inline pippy/hf/_SaveModule.py:110]: I think we want to do _save_weights in a separate function. The main reason behind it is that for the index file, we are currently just generating the index file on rank 0. However, for the weights, we definitely need to do it on every single rank.

**Comment by eddogola** [comment]
@fegin

File: .gitignore [hunk 2]
```diff
@@ -9,4 +9,5 @@ dist
 .watchmanconfig
```
Comment (wz337) [inline .gitignore:9]: You probably don't want to check in .gitignore.

File: pippy/hf/_SaveModule.py [hunk 3]
```diff
@@ -117,3 +116,25 @@ def _get_binary_filename(cur_idx: int) -> str:  # type: ignore[valid-type]
     world_size = str(dist.get_world_size()).zfill(5)
 
     return f"pytorch_model-{idx}-of-{world_size}.bin"
+
+
+def _save_checkpoint(submod: Pipe, checkpoint_dir: str) -> None:
```
Comment (wz337) [inline pippy/hf/_SaveModule.py:121]: TODO: update to reflect the correct typing here.

**Review by wz337 — APPROVED** [review]
LGTM

**Pull request merged**